### ISSUE TITLE
[6.14] rngd --list not showing full list and hangs

### ISSUE THREAD

**Issue by macmpi** [nhorman/rng-tools#170]
Using under [Alpine 3.15](https://git.alpinelinux.org/aports/tree/main/rng-tools?h=3.15-stable) on Pi/armhf @dermotbradley 
```
~$ rngd -l
Entropy sources that are available but disabled
1: TPM RNG Device (tpm)


```
Have to ctrl-c to get back to shell prompt

**Comment by nhorman** [comment]
what version of rngd are you running, and can you run it with strace to see where its stalling?

**Comment by macmpi** [comment]
it is 6.14.
Here is strace on x86_64 which shows same problem
```
execve("/usr/sbin/rngd", ["rngd", "-l"], 0x7ffda724ccd8 /* 13 vars */) = 0
arch_prctl(ARCH_SET_FS, 0x7f764711bb48) = 0
set_tid_address(0x7f764711bf90)         = 2714
brk(NULL)                               = 0x55663afd9000
brk(0x55663afdb000)                     = 0x55663afdb000
mmap(0x55663afd9000, 4096, PROT_NONE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x55663afd9000
open("/etc/ld-musl-x86_64.path", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/lib/libjitterentropy.so.3", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/local/lib/libjitterentropy.so.3", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/usr/lib/libjitterentropy.so.3", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = 3
fcntl(3, F_SETFD, FD_CLOEXEC)           = 0
fstat(3, {st_mode=S_IFREG|0755, st_size=34544, ...}) = 0
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0000#\0\0\0\0\0\0"..., 960) = 960
mmap(NULL, 40960, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f764707b000
mmap(0x7f764707d000, 16384, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED, 3, 0x2000) = 0x7f764707d000
mmap(0x7f7647081000, 8192, PROT_READ, MAP_PRIVATE|MAP_FIXED, 3, 0x6000) = 0x7f7647081000
mmap(0x7f7647083000, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED, 3, 0x7000) = 0x7f7647083000
close(3)                                = 0
open("/lib/libcrypto.so.1.1", O_RDONLY|O_LARGEFILE|O_CLOEXEC) = 3
fcntl(3, F_SETFD, FD_CLOEXEC)           = 0
fstat(3, {st_mode=S_IFREG|0755, st_size=2605744, ...}) = 0
read(3, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\0`\7\0\0\0\0\0"..., 960) = 960
mmap(NULL, 2625536, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f7646dfa000
mmap(0x7f7646e6f000, 1404928, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED, 3, 0x75000) = 0x7f7646e6f000
mmap(0x7f7646fc6000, 540672, PROT_READ, MAP_PRIVATE|MAP_FIXED, 3, 0x1cc000) = 0x7f7646fc6000
mmap(0x7f764704a000, 200704, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED, 3, 0x24f000) = 0x7f764704a000
mmap(0x7f7647077000, 16384, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f7647077000
close(3)                                = 0
mprotect(0x7f7647083000, 4096, PROT_READ) = 0
mprotect(0x7f764704a000, 176128, PROT_READ) = 0
mprotect(0x7f7647118000, 4096, PROT_READ) = 0
mprotect(0x556639779000, 4096, PROT_READ) = 0
writev(2, [{iov_base="", iov_len=0}, {iov_base="Entropy sources that are availab"..., iov_len=48}], 2) = 48
writev(2, [{iov_base="1: TPM RNG Device (tpm)\n", iov_len=24}, {iov_base=NULL, iov_len=0}], 2) = 24
open("/dev/hwrng", O_RDONLY|O_NOCTTY|O_LARGEFILE) = 3
read(3, 0x7fffe2eadd08, 16)             = -1 ENODEV (No such device)
open("/sys/devices/virtual/misc/hw_random/rng_available", O_RDONLY|O_NOCTTY|O_LARGEFILE) = 4
read(4, "\n", 16)                       = 1
close(4)                                = 0
open("/dev/urandom", O_RDONLY|O_LARGEFILE) = 4
read(4, "\204\317\2(\365\n*\30\315\243\337\0165S\334\324", 16) = 16
close(4)                                = 0
rt_sigprocmask(SIG_UNBLOCK, [RT_1 RT_2], NULL, 8) = 0
rt_sigaction(SIGUSR1, {sa_handler=0x55663976e0c4, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART, sa_restorer=0x7f76470cbc8a}, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
mmap(NULL, 28672, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f7646df3000
open("/sys/devices/system/cpu/cpu0/cache/index0/type", O_RDONLY|O_LARGEFILE) = 4
read(4, "Data\n", 10)                   = 5
close(4)                                = 0
open("/sys/devices/system/cpu/cpu0/cache/index0/size", O_RDONLY|O_LARGEFILE) = 4
read(4, "24K\n", 10)                    = 4
close(4)                                = 0
open("/sys/devices/system/cpu/cpu0/cache/index1/type", O_RDONLY|O_LARGEFILE) = 4
read(4, "Instructio", 10)               = 10
close(4)                                = 0
open("/sys/devices/system/cpu/cpu0/cache/index2/type", O_RDONLY|O_LARGEFILE) = 4
read(4, "Unified\n", 10)                = 8
close(4)                                = 0
open("/sys/devices/system/cpu/cpu0/cache/index2/size", O_RDONLY|O_LARGEFILE) = 4
read(4, "1024K\n", 10)                  = 6
close(4)                                = 0
open("/sys/devices/system/cpu/cpu0/cache/index3/type", O_RDONLY|O_LARGEFILE) = -1 ENOENT (No such file or directory)
mmap(NULL, 2097172, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f7646bf2000
open("/proc/sys/crypto/fips_enabled", O_RDONLY|O_LARGEFILE) = -1 ENOENT (No such file or directory)
madvise(0x7f7646df4000, 4096, MADV_FREE) = 0
munmap(0x7f7646df3000, 28672)           = 0
munmap(0x7f7646bf2000, 2101248)         = 0
pipe([4, 5])                            = 0
sched_getaffinity(0, 128, [0, 1, 2, 3]) = 32
sched_getaffinity(0, 8, [0, 1, 2, 3])   = 8
fcntl(5, F_SETPIPE_SZ, 99210)           = 131072
mmap(NULL, 2097172, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f7646bf9000
open("/proc/sys/crypto/fips_enabled", O_RDONLY|O_LARGEFILE) = -1 ENOENT (No such file or directory)
rt_sigprocmask(SIG_UNBLOCK, [RT_1 RT_2], NULL, 8) = 0
membarrier(MEMBARRIER_CMD_REGISTER_PRIVATE_EXPEDITED, 0) = 0
mmap(NULL, 143360, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f7646bd6000
mprotect(0x7f7646bd8000, 135168, PROT_READ|PROT_WRITE) = 0
rt_sigprocmask(SIG_BLOCK, ~[RTMIN RT_1 RT_2], [], 8) = 0
clone(child_stack=0x7f7646bf8af8, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID|0x400000, parent_tid=[2715], tls=0x7f7646bf8b38, child_tidptr=0x7f764711bf90) = 2715
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
mmap(NULL, 2097172, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f76469d0000
open("/proc/sys/crypto/fips_enabled", O_RDONLY|O_LARGEFILE) = -1 ENOENT (No such file or directory)
mmap(NULL, 143360, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f76469ad000
mprotect(0x7f76469af000, 135168, PROT_READ|PROT_WRITE) = 0
rt_sigprocmask(SIG_BLOCK, ~[RTMIN RT_1 RT_2], [], 8) = 0
clone(child_stack=0x7f76469cfaf8, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID|0x400000, parent_tid=[2716], tls=0x7f76469cfb38, child_tidptr=0x7f764711bf90) = 2716
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
futex(0x7fffe2eadca4, FUTEX_WAIT_PRIVATE, 2, NULL) = 0
mmap(NULL, 2097172, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f76467a7000
open("/proc/sys/crypto/fips_enabled", O_RDONLY|O_LARGEFILE) = -1 ENOENT (No such file or directory)
mmap(NULL, 143360, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f7646784000
mprotect(0x7f7646786000, 135168, PROT_READ|PROT_WRITE) = 0
rt_sigprocmask(SIG_BLOCK, ~[RTMIN RT_1 RT_2], [], 8) = 0
clone(child_stack=0x7f76467a6af8, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID|0x400000, parent_tid=[2717], tls=0x7f76467a6b38, child_tidptr=0x7f764711bf90) = 2717
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
mmap(NULL, 2097172, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f764657e000
open("/proc/sys/crypto/fips_enabled", O_RDONLY|O_LARGEFILE) = -1 ENOENT (No such file or directory)
mmap(NULL, 143360, PROT_NONE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f764655b000
mprotect(0x7f764655d000, 135168, PROT_READ|PROT_WRITE) = 0
rt_sigprocmask(SIG_BLOCK, ~[RTMIN RT_1 RT_2], [], 8) = 0
clone(child_stack=0x7f764657daf8, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID|0x400000, parent_tid=[2718], tls=0x7f764657db38, child_tidptr=0x7f764711bf90) = 2718
rt_sigprocmask(SIG_SETMASK, [], NULL, 8) = 0
futex(0x7fffe2eadca4, FUTEX_WAIT_PRIVATE, 2, NULL) = 0
mmap(NULL, 20480, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f7646551000
read(4, 0x55663978af80, 16)             = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGINT {si_signo=SIGINT, si_code=SI_KERNEL} ---
+++ killed by SIGINT +++

```

And 
```
# find / -name libjitterentropy*
/usr/lib/libjitterentropy.so.3.3.0
/usr/lib/libjitterentropy.so.3

```

**Comment by nhorman** [comment]
ok, i'd really you rather ran it on the arm system, if that's what you're targeting, but if we assume its the same problem, that looks like the jitterentropy source is having trouble generating entropy.  Can you confirm by running rngd -x jitter on the arm system to see if it avoids the problem

**Comment by nhorman** [comment]
also, what version of rngd ar you building?

**Comment by macmpi** [comment]
strace on armhf instance attached [output.txt](https://github.com/nhorman/rng-tools/files/8697958/output.txt)

```
$ rngd -V
rngd 6.14
Copyright 2001-2004 Jeff Garzik
Copyright 2017 Neil Horman
Copyright (c) 2001 by Philipp Rumpf
This is free software; see the source for copying conditions.  There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by nhorman** [comment]
Thank you, so that looks like a different problem than what you are seeing on your x86_64 system.  On your arm system, the jitterentropy library appears to be "stuck".  That is to say, its trying to measure the time delta between load/store operations and not observing a sufficient change in time between iterative measurements.  We can work with the jitterentropy library maintainer @smuellerDD to figure out whats going on here, but in the interim, I suggest just to get you working for now, that you disable libjitterentropy, and use the rndr entropy source instead, which IIRC pi/armhf should support.

**Comment by macmpi** [comment]
On armhf, it does same with jitter disabled in .conf file

**Comment by nhorman** [comment]
um, thats not possible. jitter is the only entropy source that uses clock_gettime in that volume, so you must not be disabling it correctly.  Additionally, rngd doesn't have a conf file of any sort.  Can you elaborate on how you are disabling the jitter source?

**Comment by macmpi** [comment]
top command shows:
` 2977     1 root     S     2872   1%   0   0% /usr/sbin/rngd -b -p /run/rngd.pid -q -x jitter -O jitter:buffer_size:4133 -O jitter:refill_thresh:4133 -O jitter:thread_count:1
`

this was [packaged](https://git.alpinelinux.org/aports/tree/main/rng-tools?h=3.15-stable) by @dermotbradley  for Alpine distribution
```
# Configuration for /etc/init.d/rngd

# Space-delimited list of entropy sources to enable.
# No need to include any source that is enabled by default.
#
# Choose from list:
#     tpm:  Is deprecated, will be removed in future version.
#
INCLUDE_ENTROPY_SOURCES=""

# Space-delimited list of entropy sources to disable.
# Useful for disabling certain entropy sources even when supported on system.
#
EXCLUDE_ENTROPY_SOURCES="jitter"


# Number of entropy bits to support, 1 <= n <= 8.
#
#ENTROPY_COUNT=8

# Time in seconds to periodically add fresh entropy to kernel entropy pool.
#
#FORCE_RESEED=300

# Device used for random number input.
#
#HWRNG_DEVICE="/dev/hwrng"

# Kernel device used for random number output.
#
#RANDOM_DEVICE="/dev/random"

# Number of bytes written to random-device at a time.
#
#STEP=64

# Do not stop feeding entropy to random device until at least this
# many bits of entropy are available in pool. Value can be
# 0 <= n <= `sysctl kernel.random.poolsize` but is usually 75% of pool size.
#
#WATERMARK=3072


# Entropy source specific options:
#
#
# darn options:
#         use_aes:(BOOLEAN)
#
#DARN_OPTIONS="use_aes:1"
#
#
# jitter options:
#         buffer_size:(INTEGER)
#         force_soft_timer:(BOOLEAN)
#         refill_thresh:(INTEGER)
#         retry_count:(INTEGER)
#         retry_delay:(INTEGER)
#         thread_count:(INTEGER)
#         use_aes:(BOOLEAN)
#
#JITTER_OPTIONS="buffer_size:16535 force_soft_timer:0 refill_thresh:16535"
#JITTER_OPTIONS="${JITTER_OPTIONS} retry_count:1 retry_delay:-1"
#JITTER_OPTIONS="${JITTER_OPTIONS} thread_count:4 use_aes:1"
#
# With the default jitter options Arm CPU-based devices typically experience
# continuous high CPU utilisation for several minutes after rngd starts.
# See https://gitlab.alpinelinux.org/alpine/aports/-/issues/13035 and
# https://github.com/nhorman/rng-tools/issues/150
# In order to address this the following configuration changes have been made:
#
#   - reduce the number of threads used to 1 in total, rather than 1 per core.
#   - reduce the buffer size from 16535 to 4133 which reduces the time taken
#     to fill the buffer (and so reduces the duration of CPU load).
#
JITTER_OPTIONS="buffer_size:4133 refill_thresh:4133 thread_count:1"
#
#
# rdrand options:
#         use_aes:(BOOLEAN)
#
#RDRAND_OPTIONS="use_aes:0"
#
#
# rndr options:
#         use_aes:(BOOLEAN)
#
#RNDR_OPTIONS="use_aes:0"


# Any extra arguments for rngd
#
EXTRA_ARGS="-q"
```

**Comment by nhorman** [comment]
ok, so alpine created a custom configuration script to set rngd options, that's fair, but I don't see the mechanism in here that translates the EXCLUDE_ENTROPY_SOURCES variable into command line options to actual do the disable.  I imagine that if you run this command:
cat /proc/`pidof rngd`/cmdline
you will see that -x jitter is not actually included in the command line.

**Comment by macmpi** [comment]
As mentioned in previous post with the `top` command report you probably under-looked, I confirm excluded source is consistent. Yet again with confimation below

```
$ cat /proc/2594/cmdline 
/usr/sbin/rngd-b-p/run/rngd.pid-q-xjitter-Ojitter:buffer_size:4133-Ojitter:refill_thresh:4133-Ojitter:thread_count:1
```

**Comment by nhorman** [comment]
well, then something is inconsistent here.  You have a command line that shows you are disabling the jitter source, but your strace clearly shows it continues to be used. 

To clarify, the strace, was that taken from pid 2594, or did you just run rngd -l (or some other command line invocation)?  based on the initial comment, I would guess the latter, which did not disable the jitter source.  if you run rngd -l -x jitter, do you still get the hang? If ts the rngd instance running as the daemon above, do you see the hang there, and if so, do you have an strace of it?

**Comment by macmpi** [comment]
Yes the latter.
Expectation was that `rngd -l` would list entropy sources of running daemon: it is not what it is intended for?
I confirm no hang on the following:
```
$ rngd -l -x jitter
Disabling 6: JITTER Entropy generator (jitter)
Entropy sources that are available but disabled
1: TPM RNG Device (tpm)
6: JITTER Entropy generator (jitter)
Available and enabled entropy sources:
Available entropy sources that failed initalization:
0: Hardware RNG Device (hwrng)
```

**Comment by nhorman** [comment]
No, you have the intention of rngd -l correct, the issue was that I was trying to reconcile what you were reporting in terms of debug information in the strace with the debug information you were reporting with the rngd process (the latter having disabled jitter, the former not).

So at this point, I think we're on the same page, your rngd daemon (executed through the init script) disables jitter and works just fine.  The rngd -l command does not disable jitter, and hangs, which in my mind synchronizes the data you've provided.

That said, what I think we're looking at here, is still a problem in which the jitterentropy library is internally reporting a stuck condition (i.e. there is not sufficient jitter in the load/store timing for rngd -l to initialize the entropy source), and so we are hanging waiting on the determination of weather to report jitter as an entropy source that is available and operational, or available but non-operational.

I think perhaps the best thing to do her is modify the jitter init routine to make its pthread_cond_wait call (which waits for jitter thread initialization), to be instead a pthread_cond_timedwait, blocking for some fixed time (say 5 seconds), and failing the initialization of jitter if it times out (implying that we don't have enough jitter to use the source).

I'll create a branch for you to test.

**Comment by nhorman** [comment]
https://github.com/nhorman/rng-tools/pull/171

I've created that PR to potentially solve your problem.  Please build that branch and let me know if rngd -l takes more than 5 seconds to complete

**Comment by macmpi** [comment]
Thanks for the quick turnaround. It may take me a bit of time to rebuild & test as I do not have a build environment handy.
I'll keep you updated. Maybe @dermotbradley will be quicker.

**Comment by dermotbradley** [comment]
Sorry for delay commenting, was tied up with other stuff.

@nhorman re: conf file, its a file specific to the Alpine-packaging that if souced by the init.d file and it used any values defined in the EXCLUDE_ENTROPY_SOURCES variable to create "-x" options for the daemon: https://git.alpinelinux.org/aports/tree/main/rng-tools/rngd-extra.initd#n22

For the previous #150 issue (that I think affected @macmpi with a single core RPI Zero?) I tweaked the config for rngd using the conf file setting JITTER_OPTIONS="buffer_size:4133 refill_thresh:4133 thread_count:1".
However I think at least part of the issue that @macmpi is now seeing with "rngd -l" is that that setting does not apply to "rngd -l", only to the rngd daemon when run via init.d - "rngd -l" still appears to initialise jitter but does not have any jitter options set.

@macmpi I'll certainly test #171, just not sure if I'll be able to reproduce the issue in the first place.

**Comment by nhorman** [comment]
I would agree with that assessment.  This patch should just let us timeout
the jitter source of it "takes too long" (5 seconds).  I'm pretty sure it
will work, I'm just a bit concerned that it will prevent jitter from
working in situations where it takes, maybe 10 seconds to generate entropy
or some such. Give it a spin when you can and let me.know.  thanks!

**Comment by macmpi** [comment]
2 comments (did not test new branch yet):
- on [`rngd -l -x jitter`](https://github.com/nhorman/rng-tools/issues/170#issuecomment-1127894690) executed on PiZero (which has hwrng), I'm curious about that part of the report: 
```
Available entropy sources that failed initalization:
0: Hardware RNG Device (hwrng)
```
- overall hang issue on `rngd -l` also applies to other platforms [as reported on x86-64](https://github.com/nhorman/rng-tools/issues/170#issuecomment-1127018674)

**Comment by nhorman** [comment]
I only preformed the rngd -l -x jitter test to reconcile the observations on the strace, and confirm that jitter was our problem here.

regarding the similar issue on the x8664 platforms, thts entirely possible.  Jitter can take a long time to operate, and hopefully this branch will catch that before affected platforms try to use it

**Comment by macmpi** [comment]
Sure, but how can I know more about this hwrng failed initialization?
This is also unexpected on that device.

**Comment by nhorman** [comment]
oh, sorry, It was too early when I read that.  you can enable debug logging (-d) to get more information about what exactly went wrong, but I almost guarantee you that device failed because it doesn't exist on your system. That is to say, you don't have a /dev/hwrng device file in your /dev directory, or you do, but its not attachedto any real rng peripheral

**Comment by macmpi** [comment]
hum... my bad: turns-out that test was probably on x86_64 at the time (did too many of them lately)

hwrng all good on Pi:
```
$ rngd -d -l -x jitter
Disabling 6: JITTER Entropy generator (jitter)
Entropy sources that are available but disabled
1: TPM RNG Device (tpm)
6: JITTER Entropy generator (jitter)
Available and enabled entropy sources:
0: Hardware RNG Device (hwrng)
Available entropy sources that failed initalization:

$ dmesg | grep rng
[    0.000000] random: get_random_u32 called from __kmem_cache_create+0x1c/0x3b8 with crng_init=1
[    2.139391] bcm2835-rng 20104000.rng: hwrng registered
[    5.934384] random: crng init done

$ ls /dev/hw*
/dev/hwrng
```

**Comment by nhorman** [comment]
copy that, let me know what your testing on that PR shows

**Comment by nhorman** [comment]
Not sure I understand, in both cases above jitter is reporting failed init.

That can happen for a number of reasons, especially on a smaller system like an rpi.  We can look into the causes in a separate issue if you like, as it seems orthogonal to this issue.

But it looks like your hang problem is resolved, so I'm going to merge 171 and we can pick up any remaining issues in a separate thread.  Thanks!

**Comment by macmpi** [comment]
> Not sure I understand, in both cases above jitter is reporting failed init.

Yes, but the 2 cases are different, and should show 2 different results I guess.
In first case, rngd daemon is NOT running.
So` sudo rngd -l `is starting and abording jitter initialization if it takes too long: all fine, and result is consistent.

In the second case, `rngd` daemon IS already running, and with `jitter` supposedly initialized as I did let it settle-down with @dermotbradley conf file parameters (checked in `top`)
One would expect in such case `sudo rngd -l ` would give information about the currently running instance (`pidof rngd`) and not the one we just issued in the command.
I'm therefore surprised that in that second case, it also says jitter is not initialized.
Makes more sense?

**Comment by macmpi** [comment]
ok, so `rngd -l` is not meant to gather information about the currently running instance: good that is clarified as I did interpret your https://github.com/nhorman/rng-tools/issues/170#issuecomment-1127919321 differently:
>No, you have the intention of rngd -l correct,

However, such usage would be beneficial to diagnose stuff I think, but obviously your call.

**Comment by macmpi** [comment]
Last bit on it.
That timeout change essentially makes sense in the context of diagnosing current running instance actually (understand which entropy source was in use, if there was some init problem, etc...): sorry if we were not on the same understanding.

Indeed, now with that timeout, it will always signal users that `jitter` fails init on less capable CPUs, by never letting it complete it's stuff: one would then conclude jitter is just not appropriate at all, based on that feedback.
You may want to add some user message related to timeout kicking-in in the command response (and eventual option not to have timeout then!...).
